code/parachain/frame/staking-rewards/src/lib.rs

```diff
@@ -774,145 +837,259 @@ pub mod pallet {
-						let old_value = stake.clone();
-						stake.stake = ratio.mul_floor(stake.stake);
-						stake.share = ratio.mul_floor(stake.share);
-						let assets: Vec<T::AssetId> = stake.reductions.keys().cloned().collect();
-						for asset in assets {
-							let reduction = stake.reductions.get_mut(&asset);
-							if let Some(value) = reduction {
-								*value = ratio.mul_floor(*value);
-							}
-						}
-						old_position_stake = stake.stake;
-						Ok(old_value)
-					},
-					None => Err(Error::<T>::StakeNotFound),
+				Stakes::<T>::try_mutate(fnft_collection_id, fnft_instance_id, |maybe_stake| {
+					let stake = maybe_stake.as_mut().ok_or(Error::<T>::StakeNotFound)?;
+
+					let old_value = stake.clone();
+					stake.stake = ratio.mul_floor(stake.stake);
+					stake.share = ratio.mul_floor(stake.share);
+
+					for (_, reduction) in &mut stake.reductions {
+						*reduction = ratio.mul_floor(*reduction);
+					}
+
+					old_position_stake = stake.stake;
+					Ok::<_, DispatchError>(old_value)
 				})?;
+
 			let left_from_one_ratio = ratio.left_from_one();
-			let assets: Vec<T::AssetId> = old_position.reductions.keys().cloned().collect();
-			for asset in assets {
-				let reduction = old_position.reductions.get_mut(&asset);
-				if let Some(value) = reduction {
-					*value = left_from_one_ratio.mul_floor(*value);
-				}
+
+			for (_, reduction) in &mut old_position.reductions {
+				*reduction = left_from_one_ratio.mul_floor(*reduction);
 			}
 
+			let new_fnft_instance_id = T::FinancialNft::get_next_nft_id(fnft_collection_id)?;
+			let old_fnft_asset_account =
+				T::FinancialNft::asset_account(fnft_collection_id, &old_position.financial_nft_id);
+			let new_fnft_asset_account =
+				T::FinancialNft::asset_account(fnft_collection_id, &new_fnft_instance_id);
+
 			let new_stake = StakeOf::<T> {
 				stake: left_from_one_ratio.mul_floor(old_position.stake),
 				share: left_from_one_ratio.mul_floor(old_position.share),
+				financial_nft_id: new_fnft_instance_id,
 				..old_position
 			};
-			let new_position = StakeCount::<T>::increment()?;
-			Stakes::<T>::insert(new_position, &new_stake);
-			// TODO (vim):
-			// 	1. Create the new financial NFT for the new position
-			//	2. transfer the split staked amount to the NFT account and lock it
-			//	3. transfer the split share amount to the NFT account and lock it
+
+			let rewards_pool = RewardPools::<T>::try_get(new_stake.reward_pool_id)
+				.map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			// Unlock, Transfer, Lock staked asset
+			T::Assets::remove_lock(
+				T::LockId::get(),
+				rewards_pool.asset_id,
+				&old_fnft_asset_account,
+			)?;
+			T::Assets::transfer(
+				rewards_pool.asset_id,
+				&old_fnft_asset_account,
+				&new_fnft_asset_account,
+				new_stake.stake,
+				keep_alive,
+			)?;
+			Self::split_lock(
+				rewards_pool.asset_id,
+				old_position.stake,
+				new_stake.stake,
+				&old_fnft_asset_account,
+				&new_fnft_asset_account,
+			)?;
```

Comment:
Collapse all of this into the split lock function to reduce redundancy, or keep it here so that we have less abstraction when transfers happen? @benluelo @vimukthi-git

Reply:
I guess it doesn't hurt to have a single function. It's just repeating the same thing anyway.

Reply:
I think it's fine like this for now - we can always refactor it later